Web Inspector: Automatically connect Web Inspector to ServiceWorker

rcaliman-apple opened the pull request:
#### a85f48918e3542ba0ef4e9ff3b0c3b32fc5f51c5
<pre>
Web Inspector: Automatically connect Web Inspector to ServiceWorker
<a href="https://bugs.webkit.org/show_bug.cgi?id=264576">https://bugs.webkit.org/show_bug.cgi?id=264576</a>
<a href="https://rdar.apple.com/114997939">rdar://114997939</a>

Reviewed by BJ Burg.

(This work was done in collaboration with Razvan and was based on his
draft at <a href="https://github.com/WebKit/WebKit/commit/377f3e10bacc5847b25176fbc2cc0e1227ea1d95.)">https://github.com/WebKit/WebKit/commit/377f3e10bacc5847b25176fbc2cc0e1227ea1d95.)</a>

This commit enables automatically inspecting and pausing the
ServiceWorkerDebuggable. The idea is similar to the same functionalities
with the JSContext/JSGlobalObjectDebuggable. The general flow is:
1. When the debuggable is first created, we optionally mark it as
   inspectable.
2. As soon as the debuggable is marked inspectable, its main thread
   (the thread that it was created on) gets blocked.
3. When the auto-launched Web Inspector frontend finishes initializing,
   it notifies the backend.
   - It&apos;s important for the debuggable to wait for this signal because
     a genuine auto-inspection must appear attached to the debuggable
     before it begins execution, respecting any breakpoints set early on
     in its script (where auto-pausing is basically a breakpoint
     before line 1).
4. The backend unpauses the blocked debuggable. If auto-pausing was
   requested, tell the debugger agent to pause.

The service worker begins executing script unless its worker thread was
specified to start in the WorkerThreadStartMode::WaitForInspector.
During that waiting period, the worker thread can perform tasks sent
into its debugging run loop, until it&apos;s signaled to stop waiting and
continue to execute the script like normal. This commit makes use of
that interface to make the service worker pause (when justified, i.e.
developerExtrasEnabled) before running the above flow resembling
auto-inspecting a JSContext.

* Source/WebCore/workers/service/context/ServiceWorkerThread.cpp:
(WebCore::threadStartModeFromSettings):
(WebCore::ServiceWorkerThread::ServiceWorkerThread):
   - When there is potentially a remote inspector that would like to
     auto-inspect, make it so that the thread waits on start before
     executing its script.

* Source/WebCore/workers/service/context/ServiceWorkerThreadProxy.h:
* Source/WebCore/workers/service/context/ServiceWorkerThreadProxy.cpp:
(WebCore::ServiceWorkerThreadProxy::ServiceWorkerThreadProxy):
(WebCore::ServiceWorkerThreadProxy::threadStartedRunningDebuggerTasks):
   - Setting inspectability is step #1 in the above flow.
   - In step #2, calling `debuggable-&gt;setInspectable(true)` might block
     already, but we don&apos;t want that until the worker thread is setup
     and have the run loop be in debug mode, so we do that in a callback
     instead.
   - In step #4, when connection to the inspector completes or fails,
     the setInspectable call only returns then, so we unblock
     the worker thread to resume code execution.

* Source/WebCore/inspector/agents/worker/WorkerDebuggerAgent.h:
* Source/WebCore/inspector/WorkerInspectorController.h:
* Source/WebCore/inspector/WorkerInspectorController.cpp:
(WebCore::WorkerInspectorController::frontendInitialized):
(WebCore::WorkerInspectorController::connectFrontend):
(WebCore::WorkerInspectorController::disconnectFrontend):
(WebCore::WorkerInspectorController::createLazyAgents):
(WebCore::WorkerInspectorController::ensureDebuggerAgent):
* Source/WebCore/workers/service/context/ServiceWorkerDebuggable.cpp:
(WebCore::ServiceWorkerDebuggable::connect):
* Source/WebCore/workers/service/context/ServiceWorkerInspectorProxy.h:
* Source/WebCore/workers/service/context/ServiceWorkerInspectorProxy.cpp:
(WebCore::ServiceWorkerInspectorProxy::connectToWorker):
   - Mimic the logic for auto-inspecting a JSContext/JSGlobalObjectDebuggable.

* Source/JavaScriptCore/inspector/protocol/Inspector.json:
   - Step #3 in the above flow, notify the backend when frontend
     completes setting up.

* Source/WebCore/workers/service/context/ServiceWorkerDebuggable.h:
  - Allow service workers to be auto-inspected. (This is checked at <a href="https://github.com/rcaliman-apple/WebKit/blob/eng/Web-Inspector-Automatically-connect-Web-Inspector-to-ServiceWorker/Source/JavaScriptCore/inspector/remote/RemoteInspectionTarget.cpp#L95)">https://github.com/rcaliman-apple/WebKit/blob/eng/Web-Inspector-Automatically-connect-Web-Inspector-to-ServiceWorker/Source/JavaScriptCore/inspector/remote/RemoteInspectionTarget.cpp#L95)</a>

* Source/WTF/wtf/PlatformEnableCocoa.h:
   - Add feature flag just in case.

Canonical link: <a href="https://commits.webkit.org/291167@main">https://commits.webkit.org/291167@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3f46c146e36a0440e11ab6126b6fa2860936af62

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/92145 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/11679 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/1245 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/97053 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/42709 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/94195 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/12009 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/20169 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/70690 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/28154 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/95146 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/9144 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/83466 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/51019 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/8863 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/1080 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/41927 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/84910 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/79182 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/1050 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/99101 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/90861 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/19248 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/14231 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/79708 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/19500 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/79329 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78963 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19560 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/23497 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/796 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/12256 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/19228 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/24404 "Failed to checkout and rebase branch from PR 20316") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/113467 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18921 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/32834 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/22379 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/20665 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/JavaScriptCore/inspector/remote/cocoa/RemoteInspectorCocoa.mm:
```diff
@@ -179,6 +179,13 @@ static bool canAccessWebInspectorMachPort()
+        NSLog(@"RemoteInspector::updateAutomaticInspectionCandidate (target)");
+        NSLog(@"m_automaticInspectionEnabled: %s", m_automaticInspectionEnabled ? "YES" : "NO");
+        NSLog(@"TARGET NAME: %s", target->name().utf8().data());
+        NSLog(@"URL: %s", target->url().utf8().data());
```
I'm not sure it's okay to log URLs?

rcaliman-apple replied:
Won't land this, see https://github.com/WebKit/WebKit/pull/20316#discussion_r1389754551

pxlcoder commented on Source/WebCore/inspector/WorkerInspectorController.cpp:
```diff
@@ -98,11 +99,40 @@ void WorkerInspectorController::workerTerminating()
+
+void WorkerInspectorController::connectFrontend(bool isAutomaticInspection, bool immediatelyPause, Function<void()>&& frontendInitializedCallback)
+{
+    WTFLogAlways("!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!");
```
I'll leave it up to other Web Inspector reviewers to comment on how much overall logging needs to be in the PR – but I think this particular statement (and the others like it) should be removed.

rcaliman-apple replied:
None of this logging is intended to land. Will remove in a future iteration. 
Left as breadcrumbs to perhaps aid reviewers: https://github.com/WebKit/WebKit/pull/20316#issuecomment-1806241303

pxlcoder commented on Source/JavaScriptCore/inspector/remote/cocoa/RemoteInspectorCocoa.mm:
```diff
@@ -346,6 +357,38 @@ static bool canAccessWebInspectorMachPort()
         pushListingsSoon();
 }
 
+NSString* RemoteInspector::debuggableTypeStringForTarget(TargetID targetIdentifier)
```
```suggestion
NSString *RemoteInspector::debuggableTypeStringForTarget(TargetID targetIdentifier)
```

Nit: Since this is an Objective-C type.

pxlcoder commented on Source/JavaScriptCore/inspector/remote/RemoteInspector.h:
```diff
@@ -195,6 +195,8 @@ class JS_EXPORT_PRIVATE RemoteInspector final
     void setupXPCConnectionIfNeeded();
     void updateFromGlobalNotifyState() WTF_REQUIRES_LOCK(m_mutex);
+
+    NSString* debuggableTypeStringForTarget(TargetID targetIdentifier);
```
```suggestion
    NSString *debuggableTypeStringForTarget(TargetID targetIdentifier);
```

the-chenergy commented on Source/WebCore/workers/service/context/ServiceWorkerThreadProxy.cpp:
```diff
@@ -84,9 +84,13 @@ ServiceWorkerThreadProxy::ServiceWorkerThreadProxy(Ref<Page>&& page, ServiceWork
+            m_remoteDebuggable->init();
+            m_remoteDebuggable->setInspectable(m_page->inspectable());
+        });
+    });
```
Doing it this way (in two callbacks) makes this run after the run loop has been setup [(here).](https://github.com/WebKit/WebKit/blob/6c209adee5ff65d632a60c344c5b4555bfbfb5df/Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.cpp#L214) I guess could also make a handler `ServiceWorkerTheradProxy::threadStarted`, but this is shorter code.

dcrousso replied:
why did the order of `init` and `setInspectable` switch?

rcaliman-apple replied:
>why did the order of init and setInspectable switch?

`init` calls into `RemoteInspector::singleton().registerTarget(this)` to generate a target id

https://github.com/WebKit/WebKit/blob/ba826db2307bf3f96accb418889ee5bcf83aac78/Source/JavaScriptCore/inspector/remote/RemoteControllableTarget.cpp#L44

`setInspectable` calls into `RemoteInspector::singleton().updateAutomaticInspectionCandidate(this)`  which expects the target id to exist to put it into `m_pausedAutomaticInspectionCandidates`

https://github.com/WebKit/WebKit/blob/f2a4e587bdfd932792b85a90e1be547904a160c4/Source/JavaScriptCore/inspector/remote/RemoteInspectionTarget.cpp#L96


`JSGlobalObject` works the same way, initializing  `JSGlobalObjectDebuggable` as soon as it's created:

https://github.com/WebKit/WebKit/blob/f2a4e587bdfd932792b85a90e1be547904a160c4/Source/JavaScriptCore/runtime/JSGlobalObject.cpp#L846

`ServiceWorkerDebuggable` used to return `automaticInspectionAllowed` as `false`. In that case, 
RemoteInspector::singleton().updateTarget(this) got called:

https://github.com/WebKit/WebKit/blob/f2a4e587bdfd932792b85a90e1be547904a160c4/Source/JavaScriptCore/inspector/remote/RemoteInspectionTarget.cpp#L98

That returned early because of the missing `targetIndetifier` and didn't post the debuggable listing to clients like Safari's Develop menu.

But the `init` called right afterwards worked just as well because it _did_ post the debuggable which was by now already marked as inspectable. It worked, but only incidentally.

Making `ServiceWorkerDebuggable` auto-inspectable means the order of calling has to be respected.

the-chenergy replied:
`m_debuggable->init()` doesn't need to be run inside a callback and could just be done in the constructor itself. I made that change. Now in practice `init` definitely runs before `setInspectable` no matter how we order the two (`init` and the callback to run `setInspectable`).

the-chenergy commented on Source/WebCore/workers/service/context/ServiceWorkerDebuggable.cpp:
```diff
@@ -67,6 +69,24 @@ void ServiceWorkerDebuggable::dispatchMessageFromRemote(String&& message)
+{
+    RemoteInspectionTarget::pauseWaitingForAutomaticInspection();
+
+    unpauseForInitializedInspector();
```
Since we make the service worker thread use `WorkerThreadStartMode::WaitForInspector` always, when there is no remote inspector or one fails to connect, that thread still needs to be unpaused manually somehow, as in those cases there is no `frontendInitialized` callback to unpause it.

dcrousso replied:
in that case what's the point of using `WorkerThreadStartMode::WaitForInspector`?

i think it might be better to only use that when we know for sure that the developer has enabled "Automatically Pause Connecting to ServiceWorkers", and then if after ~10s (or however long we want) the frontend doesnt connect we will automatically resume

im pretty sure this is how it works right now with `JSContext` automatic inspection, so id suggest we match behavior with that as much as possible

the-chenergy commented on Source/WebCore/inspector/WorkerInspectorController.cpp:
```diff
@@ -99,11 +100,31 @@ void WorkerInspectorController::workerTerminating()
+        if (m_debuggerAgent) {
+            m_debuggerAgent->enable();
+            m_debuggerAgent->pause();
+        }
```
I feel like this could be race-y. Given `JSGlobalObjectInspectorController` had to do some work to ensure its debugger agent always gets set up here, maybe I should do the same here (and introduce a similar `ensureDebuggerAgent()`).

dcrousso replied:
yeah you'd wanna force the `WorkerDebuggerAgent` to exist

dcrousso commented on Source/WebCore/inspector/agents/worker/ServiceWorkerAgent.cpp:
```diff
@@ -65,4 +66,11 @@ Inspector::Protocol::ErrorStringOr<Ref<Inspector::Protocol::ServiceWorker::Confi
 
+Inspector::Protocol::ErrorStringOr<void> ServiceWorkerAgent::initialized()
+{
+    m_environment.frontendInitialized();
```
this is the exact same code as `Inspector.initalized`, so could we just expose that to the `"service-worker"` domain/target instead of duplicating the logic inside the `ServiceWorker` domain?

the-chenergy replied:
That also works, but since we're reusing `WorkerInspectorController` rather than creating a new `ServiceWorkerInspectorController` (yet), and `WorkerInspectorController` (the `worker` debuggable type) doesn't use the `Inspector` domain, maybe this is more convenient and allows us to keep reusing the unspecialized inspector controller (without adding a `ensureInspectorAgent()` that the regular worker doesn't use).

dcrousso replied:
id far prefer using `Inspector.initialized` rather than duplicating it.  the entire point of the `Inspector` domain is to deal with things from the frontend (such as the UI being ready to handle events), so it's odd to have something along those lines be in a different domain.  `ServiceWorker.initialized` makes me think that the event is dispatched when the `ServiceWorker` is ready, not when the Web Inspector is ready, so at the very least the name needs to be changed

also, `WorkerInspectorController` already has special logic for when dealing with a `ServiceWorker` so it should be really easy to create a `InspectorAgent` only for that case

we also shouldn't need a `ensureInspectorAgent` because this should only happen after Web Inspector is opened, meaning `createLazyAgents` would've already been called

dcrousso commented on Source/WebCore/workers/service/context/ServiceWorkerInspectorProxy.cpp:
```diff
@@ -59,14 +59,17 @@ void ServiceWorkerInspectorProxy::serviceWorkerTerminated()
+    serviceWorkerThreadProxy->thread().runLoop().postDebuggerTask([isAutomaticConnection, immediatelyPause, serviceWorkerThreadProxy](ScriptExecutionContext& context) {
+        Function<void()> handleFrontendInitialized = [serviceWorkerThreadProxy] {
+            serviceWorkerThreadProxy->serviceWorkerDebuggable().unpauseForInitializedInspector();
+        };
```
is there really no way to get to the `ServiceWorkerDebuggable` from within `WorkerInspectorController`?  it seems really overkill (and annoying for object lifetime tracking) to have to pass in a lambda

NIT: id just inline this

the-chenergy replied:
I simplified this a bit.

Inlining could also work but I thought this makes it more compact that only the callback uses the debuggable, rather than having the controller hold a (weak) reference to it that's only used one time.

There seems to not be an easy way to get the debuggable from within the controller, and the controller is not specialized for service workers so maybe it makes less sense to do service-worker-specific work in there.

dcrousso commented on Source/WebCore/workers/service/context/ServiceWorkerThread.cpp:
```diff
@@ -116,8 +116,14 @@ static WorkerParameters generateWorkerParameters(const ServiceWorkerContextData&
+#if ENABLE_REMOTE_INSPECTOR
+static const WorkerThreadStartMode serviceWorkerThreadStartMode = WorkerThreadStartMode::WaitForInspector;
+#else
+static const WorkerThreadStartMode serviceWorkerThreadStartMode = WorkerThreadStartMode::Normal;
```
ideally we'd have this only happen if `DeveloperExtrasEnabled` as most users dont even have that turned on and therefore shouldn't have to "suffer" a slower startup

NIT: these could be `constexpr`

dcrousso commented on Source/WebInspectorUI/UserInterface/Protocol/Target.js:
```diff
@@ -98,6 +98,8 @@ WI.Target = class Target extends WI.Object
             if (this.hasDomain("Inspector"))
                 this.InspectorAgent.initialized();
+            else if (this.hasDomain("ServiceWorker"))
+                this.ServiceWorkerAgent.initialized();
```
the proper compatibility check here would be `this.hasCommand("ServiceWorker.initialized")` since the `ServiceWorker` domain has existed for a while whereas the `initialized` command is new

though see my above comment about how we could just expose the `Inspector` domain instead of duplicating this logic elsewhere

dcrousso reviewed (commented):
i think this is getting closer to what we want! :)

burg commented on Source/WebCore/workers/service/context/ServiceWorkerThread.cpp:
```diff
@@ -116,8 +116,18 @@ static WorkerParameters generateWorkerParameters(const ServiceWorkerContextData&
     };
 }
 
+static WorkerThreadStartMode threadStartMode(const Settings::Values& settingsValues)
```
Nit: rename to `threadStartModeFromSettings()`

burg commented on Source/WebCore/workers/service/context/ServiceWorkerThreadProxy.cpp:
```diff
@@ -101,6 +106,19 @@ ServiceWorkerThreadProxy::~ServiceWorkerThreadProxy()
     m_serviceWorkerThread->clearProxies();
 }
 
+#if ENABLE_REMOTE_INSPECTOR
```
Style: WebKit does `ENABLE(REMOTE_INSPECTOR)`

burg commented on Source/WebCore/workers/service/context/ServiceWorkerInspectorProxy.cpp:
```diff
@@ -59,15 +61,20 @@ void ServiceWorkerInspectorProxy::serviceWorkerTerminated()
-        downcast<WorkerGlobalScope>(context).inspectorController().connectFrontend();
-    });
+    serviceWorkerThreadProxy->thread().runLoop().postDebuggerTask(
+        [weakDebuggable = ThreadSafeWeakPtr { debuggable }, isAutomaticConnection, immediatelyPause, serviceWorkerThreadProxy](ScriptExecutionContext& context) {
```
I don't think serviceWorkerThreadProxy needs to be captured, it's not used? If it's for refcount protection, it should be `protectedFoo`.

cdumez commented on Source/WebCore/inspector/WorkerInspectorController.h:
```diff
@@ -54,7 +55,7 @@ class WorkerInspectorController final : public Inspector::InspectorEnvironment {
     void workerTerminating();
 
-    void connectFrontend();
+    void connectFrontend(bool isAutomaticInspection = false, bool immediatelyPause = false, Function<void()>&& frontendInitializedCallback = { });
```
We should avoid boolean parameters like these and use enumeration classes for clarity. Otherwise, the call sites can be confusing. e.g.:
```
enum class IsAutomaticInspection : bool { No, Yes };
```

cdumez commented on Source/WebCore/workers/service/context/ServiceWorkerDebuggable.h:
```diff
@@ -53,6 +53,8 @@ class ServiceWorkerDebuggable final : public Inspector::RemoteInspectionTarget {
     void disconnect(Inspector::FrontendChannel&) final;
     void dispatchMessageFromRemote(String&& message) final;
 
+    bool automaticInspectionAllowed() const final { return true; }
```
should have an `is` prefix.

dcrousso commented on Source/WebCore/inspector/WorkerInspectorController.cpp:
```diff
@@ -99,11 +99,33 @@ void WorkerInspectorController::workerTerminating()
+
+        ensureDebuggerAgent();
+        m_debuggerAgent->enable();
+        m_debuggerAgent->pause();
```
i think this would be better as
```cpp
ensureDebuggerAgent().enable();
ensureDebuggerAgent().pause();
```
to match other cases and just in case `WorkerDebuggerAgent::enable` somehow causes `m_debuggerAgent` to be cleared

dcrousso commented on Source/WebCore/inspector/WorkerInspectorController.cpp:
```diff
@@ -220,6 +243,17 @@ void WorkerInspectorController::createLazyAgents()
+void WorkerInspectorController::ensureDebuggerAgent()
+{
+    if (m_debuggerAgent)
+        return;
```
typically `ensure*` methods return a reference to the created object
```cpp
if (!m_debuggerAgent) {
    ...
}
return *m_debuggerAgent;
```

dcrousso commented on Source/WebCore/workers/service/context/ServiceWorkerThread.cpp:
```diff
@@ -116,8 +116,18 @@ static WorkerParameters generateWorkerParameters(const ServiceWorkerContextData&
+static WorkerThreadStartMode threadStartModeFromSettings(const Settings::Values& settingsValues)
+{
+#if ENABLE(REMOTE_INSPECTOR)
+    if (settingsValues.developerExtrasEnabled)
```
is there a way to also check for if the developer has enabled "Automatically Pause Connecting to ServiceWorkers"?

rcaliman-apple replied:
Not at this stage. When the service worker is spawned, a corresponding `ServiceWorkerDebuggable` is created and sent via `RemoteInspector::sendAutomaticInspectionCandidateMessage` over to the any debugger client which responds with the automatic inspection policy: ignore, inspect, inspect + pause.

the-chenergy replied:
Also, we want the debuggable to pause regardless of "Automatically Pause Connecting to ServiceWorkers" because auto-inspection without pausing still needs to happen before code execution begins.

the-chenergy replied:
I have reverted back to always using `WaitForInspector` as the start mode. I'll find a better heuristic as a follow-up task (filed [here](https://bugs.webkit.org/show_bug.cgi?id=288552), with more explanation).

dcrousso commented on Source/WebCore/workers/service/context/ServiceWorkerInspectorProxy.cpp:
```diff
@@ -59,15 +61,20 @@ void ServiceWorkerInspectorProxy::serviceWorkerTerminated()
+        [weakDebuggable = ThreadSafeWeakPtr { debuggable }, isAutomaticConnection, immediatelyPause](ScriptExecutionContext& context) {
+            Function<void()> handleFrontendInitialized = [weakDebuggable] {
+                if (RefPtr debuggable = weakDebuggable.get())
+                    debuggable->unpauseForInitializedInspector();
```
i think you might be able to get to the `ServiceWorkerDebuggable` from within the `WorkerInspectorController` via something like `m_globalScope->workerOrWorkletThread()->workerDebuggerProxy()` and then create a new `virtual void unpauseForInitializedInspector();` that does this

i just find passing in a lambda to be really brittle and confusing and hard to follow :(

the-chenergy replied:
I'll do that plumbing work in a follow-up task. There might be a cleaner way out by casting the global scope into a `ServiceWorkerGlobalScope` and getting its `thread`.

dcrousso commented on Source/WebCore/workers/service/context/ServiceWorkerThreadProxy.cpp:
```diff
@@ -101,6 +106,19 @@ ServiceWorkerThreadProxy::~ServiceWorkerThreadProxy()
 
+#if ENABLE(REMOTE_INSPECTOR)
+
+void ServiceWorkerThreadProxy::threadStartedRunningDebuggerTasks()
```
NIT: i kinda prefer this be inlined so it's a bit clearer to see the order of things

the-chenergy replied:
I didn't want the code to indent too many levels so I think I'll keep that for now.

dcrousso commented on Source/WebCore/workers/service/context/ServiceWorkerThreadProxy.cpp:
```diff
@@ -101,6 +106,19 @@ ServiceWorkerThreadProxy::~ServiceWorkerThreadProxy()
+{
+    m_remoteDebuggable->setInspectable(m_page->inspectable());
+
+    thread().runLoop().postDebuggerTask([this, protectedThis = Ref { *this }](ScriptExecutionContext&) {
```
so is the idea here that even though we're already running in a debugger task, we queue another task to stop running debugger tasks to let any previously queued debugger tasks run?  if so, definitely worth a comment explaining this.  if not, why the need for the extra `postDebuggerTask`?

the-chenergy replied:
The code that runs debugger tasks looks like this:

https://github.com/WebKit/WebKit/blob/ba826db2307bf3f96accb418889ee5bcf83aac78/Source/WebCore/workers/WorkerOrWorkletThread.cpp#L99-L101

Without sending that as an extra task, `runInDebuggerMode` inside the loop still waits for another message in order to return.

(This was always the pattern to call `stopRunningDebuggerTasks` within another queued task. Perhaps it'd be better for `stopRunningDebuggerTasks` to queue the task by itself internally, but that's the pattern we've been going with already.)

the-chenergy committed:
Web Inspector: Automatically connect Web Inspector to ServiceWorker
https://bugs.webkit.org/show_bug.cgi?id=264576
rdar://114997939

Reviewed by BJ Burg.

(This work was done in collaboration with Razvan and was based on his
draft at https://github.com/WebKit/WebKit/commit/377f3e10bacc5847b25176fbc2cc0e1227ea1d95.)

This commit enables automatically inspecting and pausing the
ServiceWorkerDebuggable. The idea is similar to the same functionalities
with the JSContext/JSGlobalObjectDebuggable. The general flow is:
1. When the debuggable is first created, we optionally mark it as
   inspectable.
2. As soon as the debuggable is marked inspectable, its main thread
   (the thread that it was created on) gets blocked.
3. When the auto-launched Web Inspector frontend finishes initializing,
   it notifies the backend.
   - It's important for the debuggable to wait for this signal because
     a genuine auto-inspection must appear attached to the debuggable
     before it begins execution, respecting any breakpoints set early on
     in its script (where auto-pausing is basically a breakpoint
     before line 1).
4. The backend unpauses the blocked debuggable. If auto-pausing was
   requested, tell the debugger agent to pause.

The service worker begins executing script unless its worker thread was
specified to start in the WorkerThreadStartMode::WaitForInspector.
During that waiting period, the worker thread can perform tasks sent
into its debugging run loop, until it's signaled to stop waiting and
continue to execute the script like normal. This commit makes use of
that interface to make the service worker pause (when justified, i.e.
developerExtrasEnabled) before running the above flow resembling
auto-inspecting a JSContext.

* Source/WebCore/workers/service/context/ServiceWorkerThread.cpp:
(WebCore::threadStartModeFromSettings):
(WebCore::ServiceWorkerThread::ServiceWorkerThread):
   - When there is potentially a remote inspector that would like to
     auto-inspect, make it so that the thread waits on start before
     executing its script.

* Source/WebCore/workers/service/context/ServiceWorkerThreadProxy.h:
* Source/WebCore/workers/service/context/ServiceWorkerThreadProxy.cpp:
(WebCore::ServiceWorkerThreadProxy::ServiceWorkerThreadProxy):
(WebCore::ServiceWorkerThreadProxy::threadStartedRunningDebuggerTasks):
   - Setting inspectability is step #1 in the above flow.
   - In step #2, calling `debuggable->setInspectable(true)` might block
     already, but we don't want that until the worker thread is setup
     and have the run loop be in debug mode, so we do that in a callback
     instead.
   - In step #4, when connection to the inspector completes or fails,
     the setInspectable call only returns then, so we unblock
     the worker thread to resume code execution.

* Source/WebCore/inspector/agents/worker/WorkerDebuggerAgent.h:
* Source/WebCore/inspector/WorkerInspectorController.h:
* Source/WebCore/inspector/WorkerInspectorController.cpp:
(WebCore::WorkerInspectorController::frontendInitialized):
(WebCore::WorkerInspectorController::connectFrontend):
(WebCore::WorkerInspectorController::disconnectFrontend):
(WebCore::WorkerInspectorController::createLazyAgents):
(WebCore::WorkerInspectorController::ensureDebuggerAgent):
* Source/WebCore/workers/service/context/ServiceWorkerDebuggable.cpp:
(WebCore::ServiceWorkerDebuggable::connect):
* Source/WebCore/workers/service/context/ServiceWorkerInspectorProxy.h:
* Source/WebCore/workers/service/context/ServiceWorkerInspectorProxy.cpp:
(WebCore::ServiceWorkerInspectorProxy::connectToWorker):
   - Mimic the logic for auto-inspecting a JSContext/JSGlobalObjectDebuggable.

* Source/JavaScriptCore/inspector/protocol/Inspector.json:
   - Step #3 in the above flow, notify the backend when frontend
     completes setting up.

* Source/WebCore/workers/service/context/ServiceWorkerDebuggable.h:
  - Allow service workers to be auto-inspected. (This is checked at https://github.com/rcaliman-apple/WebKit/blob/eng/Web-Inspector-Automatically-connect-Web-Inspector-to-ServiceWorker/Source/JavaScriptCore/inspector/remote/RemoteInspectionTarget.cpp#L95)

* Source/WTF/wtf/PlatformEnableCocoa.h:
   - Add feature flag just in case.

Canonical link: https://commits.webkit.org/291167@main
